docs: improve desktop-capturer loopback docs (#47493)
electron/electron: docs/api/desktop-capturer.md

Process: [Main](../glossary.md#main-process)

> The following example shows how to capture video and system audio (if supported) of the current desktop:

Author: I don't understand what the "title is Electron" part has to do with the example below. Let's change it to "current desktop" since that's more apt.

  <body>
    <button id="startButton" class="button">Start</button>
    <button id="stopButton" class="button">Stop</button>

>     <video width="320" height="240" autoplay controls></video>

Author: DX improvement: `controls` being visible is important for the dev to be able to easily determine if there's an audio track available or not.

[`navigator.mediaDevices.getUserMedia`]: https://developer.mozilla.org/en/docs/Web/API/MediaDevices/getUserMedia
[`systemPreferences.getMediaAccessStatus`]: system-preferences.md#systempreferencesgetmediaaccessstatusmediatype-windows-macos

> ## Loopback Audio on macOS 13/14/15

Reviewer: How about "13+"?

Author: I like that. I used "13+" in my revision.

On macOS 13 and macOS 14, `navigator.mediaDevices.getDisplayMedia` may be used to capture loopback audio if the [`setDisplayMediaRequestHandler`](./session.md#sessetdisplaymediarequesthandlerhandler-opts) callback is returned with `audio: 'loopback'`.

> For macOS 15+, loopback audio is only supported if the following experimental command line switches are appended:

Reviewer: I would say "only works if". It's an experimental feature that implies it isn't supported yet.

```js
// main.js
app.commandLine.appendSwitch('enable-features', 'MacLoopbackAudioForScreenShare,MacSckSystemAudioLoopbackOverride');
```

> [!NOTE]
> Loopback audio is not supported on any macOS version in `navigator.mediaDevices.getUserMedia`.

> 

Author: This should cover everything and (hopefully) resolve a massive amount of confusion in the community right now. :)

## Caveats

> `navigator.mediaDevices.getUserMedia` does not work on macOS for audio capture due to a fundamental limitation whereby apps that want to access the system's audio require a [signed kernel extension](https://developer.apple.com/library/archive/documentation/Security/Conceptual/System_Integrity_Protection_Guide/KernelExtensions/KernelExtensions.html). Chromium, and by extension Electron, does not provide this.

Reviewer: Should we get rid of the old docs here?

Author: I stared blankly at my screen for what had to be ten minutes, pondering wether this is still relevant or not. I think it is, but for macOS 12.3 and below—the macOS operating systems which don't support the `ScreenCaptureKit` API: https://developer.apple.com/documentation/screencapturekit/

I'll add a small paragraph at the beginning of Caveats scoping it to 12.3 or below. It's still relevant—developers may have to fallback to installing third-party drivers in their Electron app if the user is on 12.3 or below.

[`navigator.mediaDevices.getUserMedia`]: https://developer.mozilla.org/en/docs/Web/API/MediaDevices/getUserMedia
[`systemPreferences.getMediaAccessStatus`]: system-preferences.md#systempreferencesgetmediaaccessstatusmediatype-windows-macos

## System Audio Capture (macOS 13+)

 On macOS 13+, `navigator.mediaDevices.getDisplayMedia` can be used to capture system audio
 if the [`setDisplayMediaRequestHandler`](./session.md#sessetdisplaymediarequesthandlerhandler-opts) `callback` 

>  is passed a stream that has `audio: 'loopback'`. See the [example](./desktop-capturer.md) above.

Reviewer: ```suggestion
 is passed a `streams` object that has `audio: 'loopback'`. See the example above.
```

Two changes:

* "`streams` object"
* I'd remove the link because we don't have an anchor tag to link to

> [!NOTE]
> Due to ongoing changes in Chromium, capturing system audio is experimental on macOS 15+ and requires the 

> > following command line switch to be appended:

Reviewer: ```suggestion
> following command line switch to be appended. Please note that you might experience bugs.
```

We could also add a link to https://www.electronjs.org/docs/latest/api/command-line here.

>
> ```js
> // main.js
> app.commandLine.appendSwitch('enable-features', 'MacSckSystemAudioLoopbackOverride');
> ```

> ## System Audio Capture (macOS 12.3 or lower)

Reviewer: What about macOS >12.3 <=13? Should we just say "macOS 12 or lower"?

Same in the line below.

> On macOS 12.3 or lower, system audio capture is not supported due to a fundamental limitation whereby apps that want to access the system's audio require a [signed kernel extension](https://developer.apple.com/library/archive/documentation/Security/Conceptual/System_Integrity_Protection_Guide/KernelExtensions/KernelExtensions.html). Chromium, and by extension Electron, does not provide this.

Reviewer: Can we break this text between 80 and 100 characters per line?

(I know this was not broken properly before you even touched it.)

[`navigator.mediaDevices.getUserMedia`]: https://developer.mozilla.org/en/docs/Web/API/MediaDevices/getUserMedia
[`systemPreferences.getMediaAccessStatus`]: system-preferences.md#systempreferencesgetmediaaccessstatusmediatype-windows-macos

## macOS Caveats

### System Audio Capture (macOS 12.3+)

>  On macOS 12.3+, `navigator.mediaDevices.getDisplayMedia` can be used to capture system audio

Author: I'm unaware of this. 😅

 if the [`setDisplayMediaRequestHandler`](./session.md#sessetdisplaymediarequesthandlerhandler-opts) `callback` 
 is passed a `streams` object that has `audio: 'loopback'`. See the example above.

> [!NOTE]
> Due to ongoing changes in Chromium, capturing system audio is experimental on macOS 15+ and requires the 
> following [command line switch](./command-line-switches.md) to be appended. Please note that you may experience bugs.
>
> ```js
> // main.js
> app.commandLine.appendSwitch('enable-features', 'MacSckSystemAudioLoopbackOverride');

> > ```

Reviewer 2: I think this is inaccurate.

Looking at how [IsSystemLoopbackCaptureSupported()](https://source.chromium.org/chromium/chromium/src/+/main:media/audio/audio_features.cc;l=94;bpv=1;bpt=1?q=IsMacSckSystemAudioLoopbackCaptureEnabled&ss=chromium) is implemented in Chromium, loopback capture depends on both the macOS version and the state of the `MacSckSystemAudioLoopbackOverride` and `MacCatapSystemAudioLoopbackCapture` flags (which are both [disabled by default](https://source.chromium.org/chromium/chromium/src/+/main:media/audio/audio_features.cc;l=16-59;bpv=1;bpt=1?q=IsMacSckSystemAudioLoopbackCaptureEnabled&ss=chromium)). The audio system used may be based on [ScreenCaptureKit](https://developer.apple.com/documentation/screencapturekit/) or [Core Audio taps](https://developer.apple.com/documentation/coreaudio/capturing-system-audio-with-core-audio-taps).

- `version < 12.3`:
  - Loopback capture is not possible.
- `12.3 ≤ version < 13.0`:
  - Loopback capture is possible, via ScreenCaptureKit only, but is opt-in, and requires enabling the `MacSckSystemAudioLoopbackOverride` flag.
- `13.0 ≤ version < 14.2`:
  - Loopback capture is possible, via ScreenCaptureKit only, and is **enabled by default**; no flag needed.
- `14.2 ≤ version < 15.0`:
  - Loopback capture is possible, via either ScreenCaptureKit or Core Audio taps. By default, it is enabled, and done via ScreenCaptureKit. However, you can opt-in to Core Audio taps instead by enabling the `MacCatapSystemAudioLoopbackCapture` flag (which [takes precedence](https://source.chromium.org/chromium/chromium/src/+/main:media/audio/mac/audio_manager_mac.cc;l=672-681;drc=b49c254c612709bbf6efee73667c827613f1876a)).
- `version ≥ 15.0`:
  - Loopback capture is possible, via either ScreenCaptureKit or Core Audio taps. By default, it is disabled altogether.
    - To enable it via Core Audio taps (the preferred approach from macOS 15.0 and above), enable the `MacCatapSystemAudioLoopbackCapture` flag. As this flag takes precedence, it doesn't matter whether `MacSckSystemAudioLoopbackOverride` is enabled or disabled.
    - To enable it via ScreenCaptureKit (not recommended, due to [permissions prompt issues](https://source.chromium.org/chromium/chromium/src/+/main:media/audio/audio_features.cc;l=86;bpv=1;bpt=1?q=IsMacSckSystemAudioLoopbackCaptureEnabled&ss=chromium) from macOS 15.0), you must enable the `MacSckSystemAudioLoopbackOverride` flag and disable/omit the `MacCatapSystemAudioLoopbackCapture` flag.

So, in summary, to support the maximum version range, you should enable **both** `MacSckSystemAudioLoopbackOverride` (to add support for `12.3 ≤ version < 13.0`) and `MacCatapSystemAudioLoopbackCapture` (to extend support for `version ≥ 15.0`).

I should clarify that this is just based on my analysis of the latest commit on the `main` branch of Chromium. It might be that things were different in previous versions of Chromium, and it might be that Electron builds Chromium with different flags (however, I find no search results for either).

Does this match your experience?

Reviewer 2: In order to provide complete guidance, the biggest questions in my mind are:

- For `12.3 ≤ version < 13.0`: if you pass `MacSckSystemAudioLoopbackOverride`, does loopback capture work?
- For `13.0 ≤ version < 13.2`: without passing any flags, does loopback capture work at all?
- For `15.0 ≤ version`:
  - without passing any flags, does loopback capture work at all?
  - if you pass `MacCatapSystemAudioLoopbackCapture `, does loopback capture work?

Reviewer 3: Another question:

With `MacCatapSystemAudioLoopbackCapture` enable, does it mean that it only requires `System Audio Recording Only`, versus `Screen & System Audio Recording`?

Since many apps just need the audio, I'm wondering if chromium supports this?

Ofc MDN says `getDisplayMedia({ audio: true, video: false })` is [not supported](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getDisplayMedia#video) but using Core Audio Tap for audio is not a huge benefit if Screen Recording is necessary anyway.

> > ```

Reviewer: This comment sounds like it is also needed on macOS 12: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/flag_descriptions.cc;l=6247-6253;drc=19d14ae4a957ef0f158e3b37cc7f3c76e839ce5d

Reviewer 2: You can indeed try opting in to it by enabling the `MacSckSystemAudioLoopbackOverride` flag, but one downstream codepath regards ScreenCaptureKit as [too buggy](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/media/capture/screen_capture_kit_device_utils_mac.mm;l=14-20?q=ScreenCaptureKit&ss=chromium%2Fchromium%2Fsrc) to use until 13.2, so I'm not sure it'll work in practice.

### System Audio Capture (before macOS 12.3)

> On macOS 12.3 or lower, system audio capture is not supported due to a fundamental limitation whereby 

Reviewer: ```suggestion
Before macOS 12.3, system audio capture is not supported due to a fundamental limitation whereby 
```

Sounds like it works on 12.3.

Reviewer 2: You can try opting in to it by enabling the `MacSckSystemAudioLoopbackOverride` flag, but various downstream codepaths regard ScreenCaptureKit as [too buggy](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/media/capture/screen_capture_kit_device_utils_mac.mm;l=14-20?q=ScreenCaptureKit&ss=chromium%2Fchromium%2Fsrc) to use until 13.2, so I'm not sure it'll work in practice.

> apps that want to access the system's audio require a [signed kernel extension](https://developer.apple.com/library/archive/documentation/Security/ConceptualSystem_Integrity_Protection_Guide/KernelExtensions/KernelExtensions.html). 

Reviewer: ```suggestion
apps that want to access the system's audio require a
[signed kernel extension](https://developer.apple.com/library/archive/documentation/Security/ConceptualSystem_Integrity_Protection_Guide/KernelExtensions/KernelExtensions.html). 
```

Wrapping between 80 and 100 chars.
